sql: do not attempt to re-execute portal after PL/pgSQL txn control

DrewKimball opened the pull request:
Previously, the PL/pgSQL transaction control statements (COMMIT/ROLLBACK) did not work with the extended wire protocol. If the `CALL` statement was executed via portal, we would attempt to re-execute the portal after the original transaction ended. This would result in an error like `unknown portal ""`.

This commit fixes the bug by replacing the original command with a dummy `ExecStmt` command when resuming stored proc execution in a new transaction. This allows the portal to be cleaned up with the first transaction without attempting to resolve it after the fact. The fix is controled by the session var `use_proc_txn_control_extended_protocol_fix`, which is on by default.

Informs #147701

Release note (bug fix): Fixed a bug that would cause a CALL statement executed via a portal in the extended wire protocol to result in an error like `unknown portal ""` if the stored procedure contained `COMMIT` or `ROLLBACK` statements. The bug has existed since PL/pgSQL transaction control statements were introduced in v24.1.

DrewKimball committed:
sql: do not attempt to re-execute portal after PL/pgSQL txn control

Previously, the PL/pgSQL transaction control statements (COMMIT/ROLLBACK)
did not work with the extended wire protocol. If the `CALL` statement was
executed via portal, we would attempt to re-execute the portal after the
original transaction ended. This would result in an error like
`unknown portal ""`.

This commit fixes the bug by replacing the original command with a dummy
`ExecStmt` command when resuming stored proc execution in a new transaction.
This allows the portal to be cleaned up with the first transaction without
attempting to resolve it after the fact. The fix is controlled by the session
var `use_proc_txn_control_extended_protocol_fix`, which is off by default,
and will be turned on in the following commit. Only this commit should be
backported prior to v25.3.

Informs #147701

Release note (bug fix): Fixed a bug that would cause a CALL statement
executed via a portal in the extended wire protocol to result in an error
like `unknown portal ""` if the stored procedure contained `COMMIT` or
`ROLLBACK` statements. The bug has existed since PL/pgSQL transaction
control statements were introduced in v24.1. The fix will be off by
default in versions prior to v25.3.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/147923)
<!-- Reviewable:end -->

yuzefovich reviewed (approved):
Nice find! <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 9 of 9 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147923#-:-OS5_4i-5KezOarON8wX:boyyrm)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @michae2)



<!-- Sent from Reviewable.io -->

yuzefovich commented:
nit: it might be easier / cleaner to pull out a separate commit to enable the fix by default so that the first commit has it disabled for backports

michae2 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Nice job! Took me a while to understand, but I think this is right.

> Reviewed 9 of 9 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147923#-:-OS5dk5ZEHL1ZXgHPrL4:btal9ak)* status: <span>:shipit:</span> complete! 2 of 0 LGTMs obtained (waiting on @DrewKimball)

___
*[`-- commits` line 16 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/147923#-OS5WRQC5Onfrekg1EiZ:-OS5WRQC5Onfrekg1Ei_:b-7dcpyr):*
Should this say "fixes"?
___
*[`pkg/sql/pgwire/testdata/pgtest/procedure` line 65 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/147923#-OS5Y4t23SNs_Eoo4WPm:-OS5Y4t23SNs_Eoo4WPn:b-bg9isq) ([raw file](https://github.com/cockroachdb/cockroach/blob/f54b1270160ff6b9d76698fe41f1eefcd50a4191/pkg/sql/pgwire/testdata/pgtest/procedure#L65)):*
> ```
> Parse {"Name": "foo", "Query": "CALL p()"}
> Bind {"DestinationPortal": "foo", "PreparedStatement": "foo"}
> Execute {"Portal": "foo"}
> ```

Was `MaxRows` not actually needed for a repro, then?
___
*[`pkg/sql/conn_executor.go` line 1653 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/147923#-OS5c2It2Ksw-wllCMbR:-OS5c2It2Ksw-wllCMbS:b-gsed80) ([raw file](https://github.com/cockroachdb/cockroach/blob/f54b1270160ff6b9d76698fe41f1eefcd50a4191/pkg/sql/conn_executor.go#L1653)):*
> ```Go
> 
> 			// resumeStmt is set if resumeProc is set. It is used to determine the
> 			// response format for the extended wire protocol.
> ```

I don't quite understand this comment about response format. I thought the purpose was to avoid any portal-related side effects when going through the commit / rollback state machine on the way to executing the routine closure?
___
*[`pkg/sql/conn_executor.go` line 2930 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/147923#-OS5e3iOAB5BEe55vQJ2:-OS5e3iOAB5BEe55vQJ3:b-huonr8) ([raw file](https://github.com/cockroachdb/cockroach/blob/f54b1270160ff6b9d76698fe41f1eefcd50a4191/pkg/sql/conn_executor.go#L2930)):*
> ```Go
> 		Statement:    ex.extraTxnState.storedProcTxnState.resumeStmt,
> 		TimeReceived: timeReceived,
> 	}, nil
> ```

If the original command was ExecStmt, should LastInBatch be set to match the original command?
___
*[`pkg/sql/vars.go` line 4052 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/147923#-OS5dN87CWOR_PN6SEJS:-OS5dN87CWOR_PN6SEJT:bet61du) ([raw file](https://github.com/cockroachdb/cockroach/blob/f54b1270160ff6b9d76698fe41f1eefcd50a4191/pkg/sql/vars.go#L4052)):*
> ```Go
> 			return formatBoolAsPostgresSetting(evalCtx.SessionData().UseProcTxnControlExtendedProtocolFix), nil
> 		},
> 		GlobalDefault: globalTrue,
> ```

I agree with the conversation in slack, let's backport this disabled.


<!-- Sent from Reviewable.io -->

michae2 reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147923)* status: <span>:shipit:</span> complete! 2 of 0 LGTMs obtained (waiting on @DrewKimball)

___
*[`pkg/sql/conn_executor.go` line 2930 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/147923#-OS5e3iOAB5BEe55vQJ2:-OS5swjf3XsIS4ZmX0Yw:b-a0ekwa) ([raw file](https://github.com/cockroachdb/cockroach/blob/f54b1270160ff6b9d76698fe41f1eefcd50a4191/pkg/sql/conn_executor.go#L2930)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

If the original command was ExecStmt, should LastInBatch be set to match the original command?
</blockquote></details>

To broaden this question a little bit, I'm wondering if we do any work _after_ execution of CALL finishes that depends on the original ExecStmt or ExecPortal curCmd?

It seems like we only want to prevent re-doing the portal work that should semantically happen _before_ execution of CALL starts. But we still might want to do any portal work that should semantically happen _after_ execution of CALL finishes. (Maybe there is no such work, though?)

I'm wondering if wiping the entire curCmd is too much. Maybe it would be safer to keep the curCmd and instead skip over any work that semantically happens _before_ execution if storedProcTxnState is set? WDYT?


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147923)* status: <span>:shipit:</span> complete! 2 of 0 LGTMs obtained (waiting on @michae2)

___
*[`-- commits` line 16 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/147923#-OS5WRQC5Onfrekg1EiZ:-OTjs49A3AG7Q0ie47Fr:b-dfckd5):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Should this say "fixes"?
</blockquote></details>

Nope - I put some details in another comment, but this is only fixing one bug (the more urgent one) from that issue.
___
*[`pkg/sql/conn_executor.go` line 1653 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/147923#-OS5c2It2Ksw-wllCMbR:-OTjovpR8ru0rCm0x7rD:b-g91rrz) ([raw file](https://github.com/cockroachdb/cockroach/blob/f54b1270160ff6b9d76698fe41f1eefcd50a4191/pkg/sql/conn_executor.go#L1653)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

I don't quite understand this comment about response format. I thought the purpose was to avoid any portal-related side effects when going through the commit / rollback state machine on the way to executing the routine closure?
</blockquote></details>

When we re-enter the state machine with the dummy command, the connExecutor uses the statement to determine the format (e.g. count of rows affected vs rows). I'll change the comment to avoid mentioning the response format.
___
*[`pkg/sql/conn_executor.go` line 2930 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/147923#-OS5e3iOAB5BEe55vQJ2:-OTjsc5A0lz-UAxZE_9S:bsrpmls) ([raw file](https://github.com/cockroachdb/cockroach/blob/f54b1270160ff6b9d76698fe41f1eefcd50a4191/pkg/sql/conn_executor.go#L2930)):*
**TL;DR:** your second question makes sense, I'm just afraid of the connExecutor. Your first question shouldn't itself be a problem, but we need to think about whether we should be as strict as Postgres.

>To broaden this question a little bit, I'm wondering if we do any work _after_ execution of CALL finishes that depends on the original ExecStmt or ExecPortal curCmd?

I considered doing something like that, but I preferred the solution that doesn't involve modifying the existing `ExecPortal` logic to keep the surface area as small as possible. I'm open to reconsidering that if you think the other way will be safer.

>If the original command was ExecStmt, should LastInBatch be set to match the original command?

Hmm... this is definitely not needed for correctness, since `LastInBatch` is only an optimization to allow autocommit - we synthesize a `Sync` command for the simple protocol (which is where `LastInBatch` applies), which will take care of committing the transaction. 

However, there's something a bit confusing going on. For context, both Postgres and CRDB run a batch of simple-protocol statements as an implicit transaction. But Postgres doesn't allow PL/pgSQL txn control statements in that context. Here's an example:
```
cockroach % psql postgres -c 'select transaction_timestamp(); call p();'
NOTICE:  foo 2025-06-27 01:01:32.05846-06
    transaction_timestamp
------------------------------
 2025-06-27 01:01:32.05846-06
(1 row)

ERROR:  invalid transaction termination
CONTEXT:  PL/pgSQL function p() line 1 at COMMIT
```
The `psql` tool doesn't use the simple protocol, so you can run stuff like `select transaction_timestamp(); call p();` without issue - the statements just run in different (implicit) transactions.
However, our CLI _does_ use the simple protocol. We currently allow a situation like the above, where the `CALL()` is part of a simple-protocol batch. If we were to exactly match Postgres there, you wouldn't be allowed to run that batch of statements from our CLI.

Do you think we should match Postgres and return an error here? This might not be something we need to address in this PR, I suppose.
___
*[`pkg/sql/pgwire/testdata/pgtest/procedure` line 65 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/147923#-OS5Y4t23SNs_Eoo4WPm:-OTjrCAx1BqShua2knjY:b21jfys) ([raw file](https://github.com/cockroachdb/cockroach/blob/f54b1270160ff6b9d76698fe41f1eefcd50a4191/pkg/sql/pgwire/testdata/pgtest/procedure#L65)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Was `MaxRows` not actually needed for a repro, then?
</blockquote></details>

So, that issue actually tracks two bugs; the first is the `unknown portal` error that previously occurred when a stored proc was executed through a portal in the extended wire protocol. That's what this PR fixes. The node crash bug (which requires `MaxRows`) happens because pausable portals don't expect the transaction to have been committed and the portal cleaned up by the time control reaches back to the pgwire command loop. The fix for that will just be to correctly enforce the restriction described in https://github.com/cockroachdb/cockroach/issues/98911, so that pausable portals are not allowed with CALL statements.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (approved):
> Reviewed 9 of 9 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147923)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @DrewKimball and @michae2)



<!-- Sent from Reviewable.io -->

michae2 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Some questions and comments but I think this is good.

> Reviewed 9 of 9 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147923#-:-OTn8UeO1n7KoG0LFAeY:b-5pc55m)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @DrewKimball)

___
*[`pkg/sql/conn_executor.go` line 2930 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/147923#-OS5e3iOAB5BEe55vQJ2:-OTn3Z4gDPsWOS7y6ro2:b5i7t8g) ([raw file](https://github.com/cockroachdb/cockroach/blob/f54b1270160ff6b9d76698fe41f1eefcd50a4191/pkg/sql/conn_executor.go#L2930)):*
> I considered doing something like that, but I preferred the solution that doesn't involve modifying the existing `ExecPortal` logic to keep the surface area as small as possible. I'm open to reconsidering that if you think the other way will be safer.

That makes sense. I too am scared of connExecutor.

Could we save the old command in storedProcTxnState and restore it upon finishing the CALL?

> Do you think we should match Postgres and return an error here? This might not be something we need to address in this PR, I suppose.

I do think we should match PG, but it's fine with me if you want to file an issue and save it for a future PR.
___
*[`pkg/sql/conn_executor.go` line 2923 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/147923#-OTn5tbWBKzvrNuOpN8p:-OTn5tbWBKzvrNuOpN8q:b-mixlb0) ([raw file](https://github.com/cockroachdb/cockroach/blob/5dace5934c17e4c4ee47b5c68588076041927b6b/pkg/sql/conn_executor.go#L2923)):*
> ```Go
> 	var timeReceived crtime.Mono
> 	switch t := curCmd.(type) {
> 	case ExecStmt:
> ```

Why does this need to replace ExecStmt, too? I was assuming it would only have to replace ExecPortal.


<!-- Sent from Reviewable.io -->

DrewKimball committed:
sql: turn stored proc extended protocol fix on

This commit flips `use_proc_txn_control_extended_protocol_fix` on by
default. This will make the fix active in v25.3+.

Informs #147701

Release note: None

DrewKimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147923)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @michae2)

___
*[`pkg/sql/conn_executor.go` line 2930 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/147923#-OS5e3iOAB5BEe55vQJ2:-OTnGaBH1H6K9_T3OcJ9:by6l834) ([raw file](https://github.com/cockroachdb/cockroach/blob/f54b1270160ff6b9d76698fe41f1eefcd50a4191/pkg/sql/conn_executor.go#L2930)):*
> Could we save the old command in storedProcTxnState and restore it upon finishing the CALL?

We could, but I don't think we need to - the original command is still in the command buffer, since we don't advance the command while processing PL/pgSQL txn control statements. We only advance to the next command once the stored proc has finished.

>I do think we should match PG, but it's fine with me if you want to file an issue and save it for a future PR.

I opened the issue here: https://github.com/cockroachdb/cockroach/issues/149103
___
*[`pkg/sql/conn_executor.go` line 2923 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/147923#-OTn5tbWBKzvrNuOpN8p:-OTnGvXp6FHLA8sTiPAX:bfqepuu) ([raw file](https://github.com/cockroachdb/cockroach/blob/5dace5934c17e4c4ee47b5c68588076041927b6b/pkg/sql/conn_executor.go#L2923)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Why does this need to replace ExecStmt, too? I was assuming it would only have to replace ExecPortal.
</blockquote></details>

It's not strictly necessary, but it seems better to handle the commands in the same way, by synthesizing a new command to resume execution. I'll add a comment to that effect.
___
*[`pkg/sql/vars.go` line 4052 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/147923#-OS5dN87CWOR_PN6SEJS:-OTnSxvh6tmneEuKVts0:b-ofjraa) ([raw file](https://github.com/cockroachdb/cockroach/blob/f54b1270160ff6b9d76698fe41f1eefcd50a4191/pkg/sql/vars.go#L4052)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

I agree with the conversation in slack, let's backport this disabled.
</blockquote></details>

Done. I split it into two commits, with the second one flipping the var on.


<!-- Sent from Reviewable.io -->
